plugins/modules/nd_federation_member.py

```diff
@@ -0,0 +1,268 @@
+
+    # Get federation members
+    federation_member_obj = nd.query_obj(member_path, ignore_not_found_error=True).get("items")
+
+    # Query specific member
+    if clusters and state == "query" and federation_member_obj:
+        for cluster in clusters:
+            cluster_info = next(
+                (cluster_dict for cluster_dict in federation_member_obj if cluster_dict.get("spec").get("host") == cluster.get("cluster_hostname")), None
+            )
+            if cluster_info:
+                nd.existing = cluster_info
+    else:
+        nd.existing = federation_member_obj
+
+    nd.previous = nd.existing
+
+    if state == "query":
+        nd.exit_json()
+    elif state == "absent":
+        if nd.existing:
+            if not module.check_mode:
+                for member_host in federation_member_obj:
+                    if not member_host.get("spec").get("host") == local_cluster_name:
+                        cluster_member_path = "{0}/{1}".format(member_path, member_host.get("status").get("memberID"))
+                        nd.request(cluster_member_path, method="DELETE")
+
+                # Remove the federation if there are no more members.
+                if len(nd.query_obj(member_path, ignore_not_found_error=True).get("items")) == 1:
+                    if federation_info:
+                        nd.request("{0}/{1}".format(federation_path, federation_info.get("status").get("federationID")), method="DELETE")
+            nd.existing = {}
+
+    elif state == "present":
+        remove_member_list = []
+        add_member_list = []
+        payload_dict = {"DELETE": [], "POST": []}
+        if len(federation_member_obj) <= 1:
+            # if there are no members or just a local member, add all members from users.
+            add_member_list = clusters
+        else:
+
+            # Remove existing members not specified by the users.
+            for existing_member_hosts in federation_member_obj:
+                member_host = existing_member_hosts.get("spec").get("host")
+                if member_host != local_cluster_name:
+                    # Use next() to check if the member exists in user-specified clusters
+                    member_exists = next((True for user_member_host in clusters if user_member_host.get("cluster_hostname") == member_host), False)
+                    if not member_exists:
+                        remove_member_list.append(existing_member_hosts)
+
+            # Add members specified by the users.
+            for user_member_host in clusters:
+                user_host = user_member_host.get("cluster_hostname")
+                # Use next() to check if the user-specified member exists in existing members
+                found = next((True for existing_member_hosts in federation_member_obj if existing_member_hosts.get("spec").get("host") == user_host), False)
+                if not found:
+                    add_member_list.append(user_member_host)
```

Comment:
do we need the add_member_list or can we do the add request here and append to payload dict? This would prevent additional loops for in line 235

Reply:
modified

Reply:
Not sure you understood what I am trying to say, I do not see where this comment is affecting idempotency?

You would still do the request and add it to the payload dict just here ( and at the remove part ) thus the need of storing in list variables is not needed.

Reply:
Wasn't the sleep for the creation on ND site? I also do not see this in your code anymore.

Reply:
Fixed this and mode changes

Reply:
Fixed this and mode changes